Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp

```diff
@@ -0,0 +1,406 @@
+                           << Acts::VectorHelpers::phi(track.momentum()));
+      continue;
+    }
+
+    // cuts on eta
+    if (Acts::VectorHelpers::eta(track.momentum()) > m_cfg.etaMax ||
+        Acts::VectorHelpers::eta(track.momentum()) < m_cfg.etaMin) {
+      score = 0;
+      iTrack++;
+      trackScore.push_back(score);
+      ACTS_DEBUG("Track: " << iTrack
+                           << " has score = 0, due to eta cuts --- eta =  "
+                           << Acts::VectorHelpers::eta(track.momentum()));
+      continue;
+    }
+
+    // cuts on optional cuts
+    for (const auto& ambicut : optionalCuts.cuts) {
+      if (ambicut(track)) {
+        score = 0;
+        ACTS_DEBUG("Track: " << iTrack
+                             << " has score = 0, due to optional cuts.");
+        break;
+      }
+    }
+
+    if (score == 0) {
+      iTrack++;
+      trackScore.push_back(score);
+      ACTS_DEBUG("Track: " << iTrack << " score : " << score);
+      continue;
+    }
+    // Reject tracks which didn't pass the detector cuts.
+    for (std::size_t detectorId = 0; detectorId < m_cfg.detectorMap.size();
+         detectorId++) {
+      auto detector_it = m_cfg.detectorMap.find(detectorId);
+      auto detector = detector_it->second;
+
+      ACTS_DEBUG("---> Found summary information");
+      ACTS_DEBUG("---> Detector ID: " << detectorId);
+      ACTS_DEBUG("---> Number of hits: " << trackFeaturesMap[detectorId].nHits);
+      ACTS_DEBUG(
+          "---> Number of holes: " << trackFeaturesMap[detectorId].nHoles);
+      ACTS_DEBUG("---> Number of outliers: "
+                 << trackFeaturesMap[detectorId].nOutliers);
+
+      if ((trackFeaturesMap[detectorId].nHits < detector.minHits) ||
+          (trackFeaturesMap[detectorId].nHits > detector.maxHits) ||
+          (trackFeaturesMap[detectorId].nHoles > detector.maxHoles) ||
+          (trackFeaturesMap[detectorId].nOutliers > detector.maxOutliers)) {
+        score = 0;
+        ACTS_DEBUG("Track: " << iTrack
+                             << " has score = 0, due to detector cuts");
+        break;
+      }
+    }
+
+    if (score == 0) {
+      iTrack++;
+      trackScore.push_back(score);
+      ACTS_DEBUG("Track: " << iTrack << " score : " << score);
+      continue;
+    }
+
+    // real scoring starts here
+
+    if (!m_cfg.useAmbiguityFunction) {
+      // if the ambiguity function is used, the score is processed with a
+      // different algorithm than the simple score.
+      score = 100;
+      // Adding the score for each detector.
+      // detector score is determined by the number of hits/hole/outliers *
+      // hit/hole/outlier scoreWeights in a detector.
+      for (std::size_t detectorId = 0; detectorId < m_cfg.detectorMap.size();
+           detectorId++) {
+        auto detector_it = m_cfg.detectorMap.find(detectorId);
+        auto detector = detector_it->second;
+        score += trackFeaturesMap[detectorId].nHits * detector.hitsScoreWeight;
```

Comment:
What we are doing here is the following: for every track, we run on all the detector Ids. We then check that detector id in the map, if it does not exist we create it (`operator[]`). So basically we have a total of `nTracks * detectorMap.size()` entries.

Is this the expected behaviour?

Reply:
Yes this is the expected behaviors, The trackFeatureMap stores number of  hits/holes/outliers for  each track per detector, This will substitute and generalize something like npixelhits  or nTRTholes.

Reply:
I'm not fully understanding this. We are here considering all detector Ids of the entire detector, right? 
Is the entire detector contributing to the score of a single track?

Reply:
What Ragansu means is that the scoring is done separatly for each sub-detector (in ATLAS those would be the Pixel, Strip and TRT). The score weight of a pixel hole is different from a TRT one for example. This also means that almost all of the maps here are 3 or 4 elements map (<10) so them being unsorted won't make any difference

Reply:
Reading again your comment, I think the confusing point might be the name detectorID. This is NOT a volume ID, it is an arbitrary id associted with a list of volumes in the volumeMap. For the ODD there are 2 detectorID : 0 and 1 corresponding to pixel and stip

Reply:
thanks for the clarification. This is another reason to go for a simpler data structure instead of an `std::map` imho

Reply:
I don't fully disagree, a vector would probably be enough. On the other side I am not sure it will make any kind of difference in term of perf between a small map/vector the real question hers is what is the easiest to read/understand/maintain

Reply:
I personally find vectors much more understandable. Basically in the code with have M tracks and a N clusters (a maximum of N ... true). So that basically is a matrix. And accessing the elements is a constant operation since we have the indices at hand

Reply:
I have changed trackFeaturesMap to trackFeaturesVector